adapters/adocean/adocean.go

```diff
@@ -0,0 +1,322 @@
+package adocean
+
+import (
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+	"text/template"
+
+	"github.com/golang/glog"
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/macros"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const adapterVersion = "1.0.0"
+const maxUriLength = 8000
+const measurementCode = `
+	<script>
+		+function() {
+			var wu = "%s";
+			var su = "%s".replace(/\[TIMESTAMP\]/, Date.now());
+
+			if (wu && !(navigator.sendBeacon && navigator.sendBeacon(wu))) {
+				(new Image(1,1)).src = wu
+			}
+
+			if (su && !(navigator.sendBeacon && navigator.sendBeacon(su))) {
+				(new Image(1,1)).src = su
+			}
+		}();
+	</script>
+`
+
+type ResponseAdUnit struct {
+	ID       string `json:"id"`
+	CrID     string `json:"crid"`
+	Currency string `json:"currency"`
+	Price    string `json:"price"`
+	Width    string `json:"width"`
+	Height   string `json:"height"`
+	Code     string `json:"code"`
+	WinURL   string `json:"winUrl"`
+	StatsURL string `json:"statsUrl"`
+	Error    string `json:"error"`
+}
+
+func NewAdOceanBidder(client *http.Client, endpointTemplateString string) *AdOceanAdapter {
+	a := &adapters.HTTPAdapter{Client: client}
+	endpointTemplate, err := template.New("endpointTemplate").Parse(endpointTemplateString)
+	if err != nil {
+		glog.Fatal("Unable to parse endpoint template")
+		return nil
+	}
+
+	whiteSpace := regexp.MustCompile(`\s+`)
+
+	return &AdOceanAdapter{
+		http:             a,
+		endpointTemplate: *endpointTemplate,
+		measurementCode:  whiteSpace.ReplaceAllString(measurementCode, " "),
+	}
+}
+
+type AdOceanAdapter struct {
+	http             *adapters.HTTPAdapter
+	endpointTemplate template.Template
+	measurementCode  string
+}
+
+func (a *AdOceanAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "No impression in the bid request",
+		}}
+	}
+
+	consentString := ""
+	if request.User != nil {
+		var extUser openrtb_ext.ExtUser
+		if err := json.Unmarshal(request.User.Ext, &extUser); err == nil {
+			consentString = extUser.Consent
+		}
+	}
+
+	var httpRequests []*adapters.RequestData
+	for _, auction := range request.Imp {
+		newBidRequest, err := a.makeRequest(httpRequests, &auction, request, consentString)
+		if err != nil {
+			return nil, []error{err}
```

Comment:
If there's even one element of the `[]Imp` array with an error, we are discarding the rest of them. Do we want this? Or did we mean to just `continue` and try to check the rest of the array and only append the "good ones" to `httpRequests`

Reply:
I've made changes so it will be more fault tolerant.